agent/agentserver/server_test.go

```diff
@@ -276,13 +276,81 @@ func TestReadinessCheckHandler(t *testing.T) {
 	}
 }
 
+func TestReadinessCheckHandlerCache(t *testing.T) {
+	for _, tc := range []struct {
+		desc              string
+		firstCallErr      error
+		readinessCacheTTL time.Duration
+		waitInvalidation  bool
+	}{
+		{
+			desc:              "call 1 succeeds, so second call automatically succeeds",
+			firstCallErr:      nil,
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 fails, so second call performs checks",
+			firstCallErr:      errors.New("test-error"),
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 succeeds, but cache becomes invalid, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 50 * time.Millisecond,
+			waitInvalidation:  true,
+		},
+		{
+			desc:              "call 1 succeeds, but caching is disabled, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 0,
+			waitInvalidation:  false,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			require := require.New(t)
+			mocks, cleanup := newServerMocks(t)
+			defer cleanup()
+
+			mocks.sched.EXPECT().Probe().Return(tc.firstCallErr).Times(1)
+			mocks.tags.EXPECT().CheckReadiness().Return(tc.firstCallErr).Times(1)
+			mocks.ac.EXPECT().CheckReadiness().Return(tc.firstCallErr).Times(1)
+			if tc.firstCallErr != nil || tc.waitInvalidation || tc.readinessCacheTTL == 0 {
+				mocks.sched.EXPECT().Probe().Return(nil).Times(1)
+				mocks.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				mocks.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+			}
+
+			addr := mocks.startServer(Config{readinessCacheTTL: tc.readinessCacheTTL})
+			r, err := httputil.Get(fmt.Sprintf("http://%s/readiness", addr))
+			if tc.firstCallErr == nil {
+				require.Nil(err)
+				respBody, _ := ioutil.ReadAll(r.Body)
+				require.Equal("OK", string(respBody))
+			} else {
+				require.Error(err)
+			}
+
+			if tc.waitInvalidation {
+				time.Sleep(tc.readinessCacheTTL)
```

Comment:
Nitpick: Could we consider a different approach to handle waiting for invalidation? Using `time.Sleep()` consumes CPU cycles without doing anything and could also cause delays in the CI pipeline if a test is misconfigured.
For example
```golang
    {
        ...
        readinessCacheTTL: 10 * time.Minute, 
        waitInvalidation:  true,
    }
```

Reply:
I agree, I've reworked the test to avoid sleeping.